File: src/lib/trending.ts
import type { Recipe } from "./types";

/**
 * Hacker-News-style time-decayed heat score. Cooking a recipe is the
 * strongest signal (someone actually made it), then comments, then votes.
 */
export function trendingScore(recipe: Recipe, now = Date.now()): number {
  const hours = Math.max(
    0,
    (now - new Date(recipe.created_at).getTime()) / 3_600_000,
  );
  const heat =
    recipe.net_upvotes + 3 * recipe.cook_count + 2 * recipe.comment_count + 1;
  return heat / Math.pow(hours + 2, 1.4);
}

export function sortByTrending(rows: Recipe[]): Recipe[] {
  const now = Date.now();
  return [...rows].sort((a, b) => trendingScore(b, now) - trendingScore(a, now));
}

File: ios/Adaptable/Adaptable/Utilities/Trending.swift
import Foundation

/// Hacker-News-style time-decayed heat score. Cooking a recipe is the
/// strongest signal (someone actually made it), then comments, then votes.
/// Mirrors `src/lib/trending.ts`.
enum Trending {
    static func score(_ recipe: Recipe, now: Date = Date()) -> Double {
        let created = ISODate.parse(recipe.created_at ?? "") ?? now
        let hours = max(0, now.timeIntervalSince(created) / 3600)
        let upvotes = recipe.net_upvotes ?? 0
        let cooks = recipe.cook_count ?? 0
        let comments = recipe.comment_count ?? 0
        let heat = Double(upvotes + 3 * cooks + 2 * comments + 1)
        return heat / pow(hours + 2, 1.4)
    }

    static func sorted(_ rows: [Recipe]) -> [Recipe] {
        let now = Date()
        return rows.sorted { score($0, now: now) > score($1, now: now) }
    }
}
